pgdog/src/backend/pool/dns_cache.rs

```diff
@@ -0,0 +1,362 @@
+use futures::future::join_all;
+use hickory_resolver::{name_server::TokioConnectionProvider, ResolveError, Resolver};
+use once_cell::sync::Lazy;
+use parking_lot::RwLock;
+use std::collections::{HashMap, HashSet};
+use std::net::IpAddr;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::time::sleep;
+use tracing::{debug, error};
+
+use crate::backend::Error;
+use crate::config::config;
+
+// -------------------------------------------------------------------------------------------------
+// ------ Singleton(s) -----------------------------------------------------------------------------
+
+static DNS_CACHE: Lazy<Arc<DnsCache>> = Lazy::new(|| Arc::new(DnsCache::new()));
+
+// -------------------------------------------------------------------------------------------------
+// ------ DnsCache :: Public interface -------------------------------------------------------------
+
+pub struct DnsCache {
+    resolver: Arc<Resolver<TokioConnectionProvider>>,
+    cache: Arc<RwLock<HashMap<String, IpAddr>>>,
+    hostnames: Arc<RwLock<HashSet<String>>>,
+}
+
+impl DnsCache {
+    /// Retrieve the global DNS cache instance.
+    pub fn global() -> Arc<DnsCache> {
+        DNS_CACHE.clone()
+    }
+
+    /// Create a new DNS cache instance.
+    pub fn new() -> Self {
+        // Initialize the Resolver with system config (e.g., /etc/resolv.conf on Unix)
+        let resolver = Resolver::builder(TokioConnectionProvider::default())
+            .unwrap()
+            .build();
+
+        DnsCache {
+            resolver: Arc::new(resolver),
+            cache: Arc::new(RwLock::new(HashMap::new())),
+            hostnames: Arc::new(RwLock::new(HashSet::new())),
+        }
+    }
+
+    /// Resolve hostname to socket address string.
+    pub async fn resolve(&self, hostname: &str) -> Result<IpAddr, Error> {
+        if let Some(ip) = self.get_cached_ip(hostname) {
+            return Ok(ip);
+        }
+
+        // Track hostname for future refreshes.
+        {
+            let mut hostnames = self.hostnames.write();
+            if hostnames.insert(hostname.to_string()) {
+                debug!("dns cache added hostname \"{}\"", hostname);
+            }
+        }
+
+        let ip = self.resolve_and_cache(hostname).await?;
+        Ok(ip)
+    }
+
+    /// Start the background refresh loop.
+    /// This spawns an infinite loop that runs until the process exits.
+    /// For testing, use short TTLs and allow the test runtime to handle shutdown.
+    pub fn start_refresh_loop(self: &Arc<Self>) {
+        let interval = Self::ttl();
```

Comment:
Fetch this inside the loop. Otherwise, the refresh loop interval will not be changeable at runtime. We want this setting to be dynamically changeable, e,g.:

1. Change the config
2. Run `RELOAD` in the admin DB (or send `SIGHUP` to pgdog to reload config)